[meshcop] support runtime custom vendor and product name

wgtdkp opened the pull request:
Existing implementation will publish the _meshcop._udp service
immediately when the program is started, even when there is no
provisioned Thread network. This works as expected as the Thread spec
defines.
    
The problem is that for non-Android linux platforms, the vendor,
product name are customized via build time flags but customization needs
to be runtime for Android because all Android device shares the same
Thread stack in the mainline module.
    
To support Android, we make updates to allow delaying the advertisement
of _meshcop._udp service after the customized vendor/product name has
been provided via the new "OTBR_STOP_BORDER_AGENT_ON_INIT" build flag
and the "BorderAgent::SetMeshcopServiceValues()" method.
    
Check https://android-review.git.corp.google.com/c/platform/packages/modules/Connectivity/+/2993930
for how this change is integrated with Android.

superwhd commented on src/border_agent/border_agent.cpp:
```diff
@@ -143,7 +141,39 @@ BorderAgent::BorderAgent(otbr::Ncp::ControllerOpenThread &aNcp, Mdns::Publisher
+    setVendorAndProductName(OTBR_VENDOR_NAME, OTBR_PRODUCT_NAME);
+}
+
+void BorderAgent::setVendorAndProductName(const std::string &aVendorName, const std::string &aProductName)
```
Should we also trigger `UpdateMeshCopService`?

wgtdkp replied:
This method is supposed to be called only before Thread is enabled. See the documentation of this API

superwhd commented on src/border_agent/border_agent.cpp:
I think you also need to update line 393~394 using the `mVendorName` and `mProductName`.

wgtdkp replied:
Good catch, updated

wgtdkp commented on src/agent/application.cpp:
```diff
@@ -91,30 +91,39 @@ Application::Application(const std::string               &aInterfaceName,
+{
+    if (mThreadEnabled != aThreadEnabled)
+    {
+        mThreadEnabled = aThreadEnabled;
```
Publisher is Thread-independent.

Added BackboneAgent handler here.

wgtdkp commented on src/border_agent/border_agent.cpp:
```diff
@@ -143,7 +141,39 @@ BorderAgent::BorderAgent(otbr::Ncp::ControllerOpenThread &aNcp, Mdns::Publisher
+        mProductName = aProductName;
+    }
+
+#ifdef OTBR_MESHCOP_SERVICE_INSTANCE_NAME
```
Cmake users are supposed to use the `OTBR_MESHCOP_SERVICE_INSTANCE_NAME` build option to customize the instance name. See https://github.com/openthread/ot-br-posix/blob/58577e5eb19a71c0842d8b5f62a4a30620fc2070/CMakeLists.txt#L40

superwhd commented on src/border_agent/border_agent.hpp:
```diff
@@ -94,6 +96,24 @@ class BorderAgent : private NonCopyable
+     *                          and an empty string will be ignored
+     * @param[in] aVendorName   The vendor name; must not exceed length of kMaxVendorNameLength
+     *                          and an empty string will be ignored
+     * @param[in] aVendorOui    The vendor OUI; must has length of 3 bytes or be empty and ignored
```
```suggestion
     * @param[in] aVendorOui    The vendor OUI; must have length of 3 bytes or be empty and ignored
```

superwhd commented on src/border_agent/border_agent.cpp:
```diff
@@ -504,7 +543,7 @@ std::string BorderAgent::GetAlternativeServiceInstanceName() const
     std::stringstream                       ss;
 
-    ss << BaseServiceInstanceName() << " (" << rand << ")";
+    ss << mServiceInstanceName << " (" << rand << ")";
```
Note that `GetAlternativeServiceInstanceName()` could be called multiple times when generating the alternative name. This could result in appending the random suffix multiple times to the instance name.
```
// on conflict:
mServiceInstanceName = GetAlternativeServiceInstanceName();
```

superwhd commented on src/agent/main.cpp:
```diff
@@ -253,6 +256,17 @@ static int realmain(int argc, char *argv[])
+        case OTBR_OPT_THREAD_ENABLED:
+            if (optarg == nullptr)
+            {
+                threadEnabled = true;
```
Shall we name it as `enableThread` to be consistent with `enableAutoAttach`?

superwhd commented on src/agent/application.hpp:
```diff
@@ -109,15 +109,23 @@ class Application : private NonCopyable
+     *
+     * @param[in] aThreadEnabled  A boolean indicates whether Thread is enabled or not
+     */
+    void onThreadEnabledChanged(bool aThreadEnabled);
```
```suggestion
    void OnThreadEnabledChanged(bool aThreadEnabled);
```

superwhd commented on src/border_agent/border_agent.cpp:
```diff
@@ -484,12 +516,19 @@ bool BorderAgent::IsThreadStarted(void) const
+    std::string         instanceName = mVendorName + " " + mProductName;
+
+    // The service instance name can always be overriden by `OTBR_MESHCOP_SERVICE_INSTANCE_NAME`
+    if (strcmp(OTBR_MESHCOP_SERVICE_INSTANCE_NAME, DEFAULT_MESHCOP_SERVICE_INSTANCE_NAME) != 0)
```
Do we really need `DEFAULT_MESHCOP_SERVICE_INSTANCE_NAME`? What about we just set the `instanceName` as `OTBR_MESHCOP_SERVICE_INSTANCE_NAME`?

superwhd commented on src/border_agent/border_agent.cpp:
```diff
@@ -143,7 +139,50 @@ BorderAgent::BorderAgent(otbr::Ncp::ControllerOpenThread &aNcp, Mdns::Publisher
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    VerifyOrExit(aProductName.size() <= kMaxProductNameLength, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aVendorName.size() <= kMaxVendorNameLength, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(aVendorOui.empty() || aVendorOui.size() == kVendorOuiLength, error = OTBR_ERROR_INVALID_ARGS);
+
+    mProductName = aProductName;
+    mVendorName  = aVendorName;
+    mVendorOui   = aVendorOui;
+
```
Why not reuse `SetPredefinedVendorTxtValues`?

wgtdkp replied:
SetPredefinedVendorTxtValues is removed, it's a rebase error. Sorry for that

superwhd commented on src/border_agent/border_agent.cpp:
```diff
@@ -484,12 +519,13 @@ bool BorderAgent::IsThreadStarted(void) const
 {
     const otExtAddress *extAddress = otLinkGetExtendedAddress(mNcp.GetInstance());
     std::stringstream   ss;
+    std::string         instanceName = mVendorName + " " + mProductName;
```
This seems unnecessary.

superwhd reviewed (approved):
LGTM 👍

bukepo reviewed (approved):
Not sure whether these info can be retrieved from Android platform.

abtink commented on src/agent/application.hpp:
```diff
@@ -109,15 +109,23 @@ class Application : private NonCopyable
+     *
+     * @param[in] aThreadEnabled  A boolean indicates whether Thread is enabled or not
+     */
+    void OnThreadEnabledChanged(bool aThreadEnabled);
```
`OnThreadEnabledChanged()` is named as if signaling an event with `aThreadEnabled` input. 
- The  `aThreadEnabled` name itself may a bit confusing.  
- E.g., we have `otThreadSetEnabled()` which I think is not related to this.

If I understand the intention, the idea here is for next layer code (e.g. Android) to use this after it has set vendor/product name so we can start advertising "meshcop" service in AIL (i.e. we are ready to enable the "meshcop" agent).

How about renaming this to make it clear? Maybe `SetMeshcopAgentEnabled()`?

wgtdkp replied:
Yeah, the Android level Thread enabled state is a bit different with `otThreadSetEnabled()`: When Thread is disabled in Android, apps are not allowed to call majority of the Android Thread APIs (i.e. apps are returned error code ERROR_THREAD_DISABLED), while some of the APIs which don't operate over radio are still allowed.

abtink commented on src/agent/application.cpp:
```diff
@@ -91,29 +91,45 @@ Application::Application(const std::string               &aInterfaceName,
 #endif
 #if OTBR_ENABLE_DNSSD_DISCOVERY_PROXY
-    mDiscoveryProxy.SetEnabled(true);
+    mDiscoveryProxy.SetEnabled(mThreadEnabled);
```
Is it safe/okay to possibly delay the call to `SetEnabled()` for `mAdvertisingProxy`, `mDiscoveryProxy` and `mBackboneAgent`?

Can we have the OT core stack trying to interact with these platform modules and cause issue if they are not enabled?

abtink replied:
Thinking more on this: how about we limit this only for `mBorderAgent.SetEnabled()` (which I think the main use-case here)? And still enable other components from `Init()`?

abtink commented:
> The problem is that for non-Android linux platforms, the vendor,
> product name are customized via build time flags but customization needs
> to be runtime for Android because all Android device shares the same
> Thread stack in the mainline module.

An idea for a possibly simpler way.
- We can change `BorderAgent` so that it waits until `SetMeshcopServiceValues()` is called before it will publish the "meshcop" service (i.e. add checks in `PublishMeshCopService()` for this).
- We can have this behavior be defined under a build config flag (so we can enable it for Andriod but not other platform)
- This way we don't need to add a new method in `Application` to delay the call to `BorderAgent::SetEnabled()`. The `SetMeshcopServiceValues()` call itself can act as the trigger.

Thoughts?

wgtdkp commented:
> > The problem is that for non-Android linux platforms, the vendor,
> > product name are customized via build time flags but customization needs
> > to be runtime for Android because all Android device shares the same
> > Thread stack in the mainline module.
> 
> An idea for a possibly simpler way.
> 
> * We can change `BorderAgent` so that it waits until `SetMeshcopServiceValues()` is called before it will publish the "meshcop" service (i.e. add checks in `PublishMeshCopService()` for this).
> * We can have this behavior be defined under a build config flag (so we can enable it for Andriod but not other platform)
> * This way we don't need to add a new method in `Application` to delay the call to `BorderAgent::SetEnabled()`. The `SetMeshcopServiceValues()` call itself can act as the trigger.
> 
> Thoughts?

@abtink Thanks for thoughtful review! Though my initial intention is for supporting runtime enablement of more components, I changed this CL to focus on BorderAgent for simplicity (I added a build flag OTBR_START_BORDER_AGENT_WHEN_INIT). PTAL

abtink reviewed (approved):
LGTM. Thanks @wgtdkp  for changing the code. 👍

wgtdkp commented on src/border_agent/border_agent.hpp:
```diff
@@ -94,6 +94,27 @@ class BorderAgent : private NonCopyable
+    /**
+     * Overrides MeshCoP service (i.e. _meshcop._udp) instance name, product name , vendor name and vendor OUI.
+     *
+     * This method should be called before this BorderAgent is enabled by SetEnabled.
```
Updated to "must"
